Detectors/EMCAL/simulation/src/DigitsWriteoutBuffer.cxx

```diff
@@ -20,62 +20,104 @@
+    iNode.mRecordMode = false;
+    iNode.mEndWindow = false;
+    iNode.mDigitMap->clear();
+  }
+
+  mTimedDigitsPast.clear();
 }
 
 void DigitsWriteoutBuffer::reserve()
 {
-  //Fill in the missing entries
-  for (int itime = 0; itime < (mBufferSize - mTimedDigits.size()); itime++) {
-    mTimedDigits.push_back(std::unordered_map<int, std::list<LabeledDigit>>());
+  for (int itime = 0; itime < mBufferSize; itime++) {
+    mTimedDigitsFuture.push_back(DigitTimebin());
   }
 }
 
-void DigitsWriteoutBuffer::addDigit(unsigned int towerID, LabeledDigit dig, double eventTime)
+// Add digits to the buffer
+void DigitsWriteoutBuffer::addDigits(unsigned int towerID, std::vector<LabeledDigit> digList)
 {
-  int nsamples = int(eventTime / mTimeBinWidth);
 
-  if (nsamples >= mTimedDigits.size()) {
-    mTimedDigits.pop_front();
-    mTimedDigits.push_back(std::unordered_map<int, std::list<LabeledDigit>>());
-    nsamples = mTimedDigits.size() - 1;
-  }
+  for (int ientry = 0; ientry < mBufferSize; ientry++) {
 
-  auto& timeEntry = mTimedDigits[nsamples];
+    auto& buffEntry = mTimedDigitsFuture[ientry];
+    auto dig = digList.at(ientry);
 
-  auto towerEntry = timeEntry.find(towerID);
-  if (towerEntry == timeEntry.end()) {
-    towerEntry = timeEntry.insert(std::pair<int, std::list<o2::emcal::LabeledDigit>>(towerID, std::list<o2::emcal::LabeledDigit>())).first;
+    auto towerEntry = buffEntry.mDigitMap->find(towerID);
+    if (towerEntry == buffEntry.mDigitMap->end()) {
+      towerEntry = buffEntry.mDigitMap->insert(std::pair<int, std::list<o2::emcal::LabeledDigit>>(towerID, std::list<o2::emcal::LabeledDigit>())).first;
+    }
+
+    buffEntry.mTimestamp = dig.getTimeStamp();
+
+    // if the digit time is equal to the mLiveTime, the time bin is set to be the end time window
+    // if the digit time is between mLiveTime and busy time the time bin is set to not record mode
+    // if the digit time is > (mLiveTime + mBusyTime - mPreTriggerTime), the time bin will be set to record mode
+    if (dig.getTimeStamp() == mLiveTime) {
+      buffEntry.mRecordMode = false;
+      buffEntry.mEndWindow = true;
+    } else if (dig.getTimeStamp() >= mLiveTime && dig.getTimeStamp() < (mLiveTime + mBusyTime - mPreTriggerTime)) {
+      buffEntry.mRecordMode = false;
+      buffEntry.mEndWindow = false;
+    } else if (dig.getTimeStamp() < mLiveTime || dig.getTimeStamp() >= (mLiveTime + mBusyTime - mPreTriggerTime)) {
+      buffEntry.mRecordMode = true;
+      buffEntry.mEndWindow = false;
+    }
+
+    towerEntry->second.push_back(dig);
   }
+}
 
-  towerEntry->second.push_back(dig);
+// When the current time is forwarded (every 100 ns) the entry 0 of the future dequeue
+// is removed and pushed at the end of the past dequeue. At the same time a new entry in
+// the future dequeue is pushed at the end, and - in case the past dequeue reached 15 entries
+// the first entry of the past dequeue is removed.
+void DigitsWriteoutBuffer::forwardMarker(double eventTime)
```

Comment:
I was thinking to handle the live/busy mode here: we would add two more parameters first to check whether with the event we open the readout (600 ns before end of busy), this would set the digit timebins in the future dequeue to record mode, and then a second option for the trigger, which would  the last entry in the future dequeue to "close window". The you could get rid of the time window handling in the addDigit function.

Reply:
Done